### DISCUSSION TITLE
How can you use Preact components inside of a React project?

### DISCUSSION THREAD

**Discussion by adamzerner** [preactjs/preact#4236] [Q&A] [answered]
I am currently working on [RFUI](https://github.com/adamzerner/rfui), a component library similar to [Chakra UI](https://chakra-ui.com/). I love Deno's web framework, [Fresh](https://fresh.deno.dev/), which uses Preact. I plan on using it myself on my projects for the foreseeable future, and I'd like RFUI to work well with it.

However, I'd also like it to work with people who are using React, as that's a huge audience. I'm not sure how that'd work though. Looking at [this blog post](https://dzone.com/articles/seamlessly-render-a-preact-component-in-a-react-pr), it seems tricky.

> Where they differ greatly is what that `<h1>I am a header</h1>` compiles to. For React, it's a `createElement()` call. For Preact, it's a `h()` call.

Are there any guides or documentation for how to approach this sort of situation?

(This seems like an important question. I expect that people would be much more motivated to write Preact-first components if it was clearly explained how people in React projects would be able to use those components. Although maybe that is misguided. Maybe it isn't actually important to be able to write Preact-first components since Preact makes it easy to use React-first components.)

**Comment by marvinhagemeister** [comment, marked as answer]
That blog post from 2018 about the different JSX factories is outdated. Since we launched Preact 10.x in 2019 Preact ships with both the `h()` and `createElement()` factory out of the box. The main difference when authoring component libraries for React or Preact are the imports. For React you'll be importing from `react` and for Preact that will be `preact` or `preact/hooks`. Most folks who dual publish for React and Preact publish separate npm packages where the imports are rewritten accordingly. That's pretty much all there is to it.

**Reply by adamzerner** [reply]
Oh, awesome! Thanks Marvin.

**Reply by rschristian** [reply]
Someone asked me for a demo of this a while back so I wrote up this: https://github.com/rschristian/react-preact-lib

Shows off separate builds for React and Preact in the same package, rather than separate, just as another option if you didn't want to split them out.

**Reply by adamzerner** [reply]
@rschristian Thanks, I appreciate that. I'll take a look.

**Reply by adamzerner** [reply]
@marvinhagemeister I am running into an issue actually.

For my Preact components I have stuff like `JSX.HTMLAttributes<HTMLDivElement>` for my types. But `JSX` comes from Preact, so for React components I'll need a different type.

Looking at stuff like [this](https://stackoverflow.com/q/55892409/1927876), it looks like `React.HTMLProps<HTMLDivElement>` would be the way to go. However, the following doesn't work.

```
import React from "react";

const Example = ({ ...rest }: React.HTMLProps<HTMLDivElement>) => {
  return <div {...rest}>Example</div>;
};
```

It leads to the following error:

```
Type '{ children: string; accept?: string | undefined; acceptCharset?: string | undefined; action?: string | undefined; allowFullScreen?: boolean | undefined; allowTransparency?: boolean | undefined; ... 362 more ...; key?: Key | ... 1 more ... | undefined; }' is not assignable to type 'HTMLAttributes<HTMLDivElement>'.
  Types of property 'contentEditable' are incompatible.
    Type '"plaintext-only" | Booleanish | "inherit" | undefined' is not assignable to type '"" | Booleanish | "plaintext-only" | SignalLike<"" | Booleanish | "plaintext-only" | undefined> | undefined'.deno-ts(2322)
Type '{ children: string; accept?: string | undefined; acceptCharset?: string | undefined; action?: string | undefined; allowFullScreen?: boolean | undefined; allowTransparency?: boolean | undefined; ... 362 more ...; key?: Key | ... 1 more ... | undefined; }' is not assignable to type 'HTMLAttributes<HTMLDivElement>'.
  Types of property 'contentEditable' are incompatible.
    Type '"plaintext-only" | Booleanish | "inherit" | undefined' is not assignable to type '"" | Booleanish | "plaintext-only" | SignalLike<"" | Booleanish | "plaintext-only" | undefined> | undefined'.
      Type '"inherit"' is not assignable to type '"" | Booleanish | "plaintext-only" | SignalLike<"" | Booleanish | "plaintext-only" | undefined> | undefined'.deno-ts(2322)
```

At first I thought this wasn't a Fresh-specific problem, but after asking [this question](https://stackoverflow.com/q/77719713/1927876), I have come to suspect it is. That error doesn't happen in a Remix app or on StackBlitz.

**Reply by marvinhagemeister** [reply]
Good catch, looks like we're missing the `"inherit"` value in our types. Made a PR https://github.com/preactjs/preact/pull/4242

**Reply by onordgren** [reply]
Hey! Any recommendations on how to best approach this for a typescript project? We currently have a component library written in preact that we would like to make useable from a react project. We are however running into issues when it comes to stuff like `class` vs `className` and how children are defined.

**Reply by rschristian** [reply]
You'll have to use the React-compatible props, there's no compat layer going the other way. Preact supports `className` natively so using it shouldn't be an issue.

**Reply by onordgren** [reply]
Ok, will gave that a go, thx @rschristian!
